Merge feature/renewable into GUI

flomnes opened the pull request.

guilpier-code committed:
Renewable clusters main panel : first work

Adding notebook "Renewable" tab

Inside the notebook :
- Adding area selector
- Adding a "Summary for the area" and a "Renewable cluster list" buttons.
- Adding buttons for : cluster creation, removal and clone

None of them do what it is supposed to do at this stage.

guilpier-code committed:
Renewable clusters main panel : creation of a cluster

guilpier-code committed:
Renewable clusters main panel : allowing to remove a cluster

The cluster is also removed from the area's list.ini

guilpier-code committed:
Renewable clusters main panel : adding property frame - first draft

with this commit, we can :
- Display the renewable cluster property frame
- Change some of the properties of a cluster in the property frame (by clicking first on the cluster in list of clusters frame) :
  A change of name, group or enable status is broadcast to the list of cluster frame.

guilpier-code committed:
Renewable clusters main panel : adding property frame - correcting synchro with inspector

Before that commit, clicking on a cluster inside an area's cluster list panel did not cause inspector refresh.
Now it does.

guilpier-code committed:
Renewable clusters main panel : adding area's summary

guilpier-code committed:
Renewable clusters main panel : nb of clusters in area, on areas selector.

In renewable clusters tab, on areas selector, to each area should be associated a number of clusters the area contains.
This the case for thermal clusters.
This commit make that possible for renewable clusters.

guilpier-code committed:
Renewable clusters main panel : only comments on what to not forget to do

guilpier-code committed:
Renewable clusters main panel : adding modulation tab

guilpier-code committed:
Renewable clusters main panel : adding time series tab

guilpier-code committed:
GUI & renewable clusters : adding scenario builder

guilpier-code committed:
UI time-series management : adding the switch between old and new renewables

guilpier-code committed:
UI time-series management : moving ts management grid creation

This is originally a test to see if ts management grid creation could be moved from constructor to another method
where we are sure that the study is loaded.
It will be useful for next development, so we keep it as a commit.

guilpier-code committed:
UI time-series management : renaming TS management renderer files with more proper names

guilpier-code committed:
UI time-series management : renaming class for the TS management renderer

Old name of class "SimulationTSManagement" is changed into "TSmanagement", shorter.

guilpier-code committed:
UI time-series management : correcting a previous commit (files renaming)

guilpier-code committed:
UI time-series management : time series management conditional display

Displaying a different time series management grid depending the value of advanced parameter : renewable generation modeling
- if it is set to "aggregated", the grid is the usual one
- if it is set to "clusters", the Wind and Solar usual columns are replaced with a "Renewable" column (for clusters)

guilpier-code committed:
Merge remote-tracking branch 'remotes/origin/feature/renewable' into feature/ui-time-series-management

guilpier-code committed:
Update UI from feature/renewable : make compilation succeed

Previous merge was done automatically by git algorithms.
After this merge, some work has to be done so that compilation succeeds.
This commit performs it.

flomnes committed:
Merge remote-tracking branch 'origin/feature/renewable' into HEAD

Peer merging with gp

flomnes commented on src/libs/antares/study/parameters.cpp:
```diff
@@ -741,6 +742,20 @@ static bool SGDIntLoadFamily_R(Parameters& d, const String& key, const String& v
+        }
+        logs.warning() << "parameters: invalid renewable generation modelling. Got '" << value
+            << "'. reset to aggregated";
+        d.renewableGeneration.rgModelling = rgAggregated;
```
These 2 lines are not useful : the program exits since the function returns `false`.

Also, I find this syntax easier to read
```cpp
return ConvertCStrToListTimeSeries(value, d.timeSeriesToRefresh);
```

flomnes replied:
Change 6ec61772 belongs in branch feature/renewable. I cherry-picked the backend part into PR #283. I will update this PR with the appropriate solver/TS generator behavior.

No change required for now, but expect minor merge conflicts later.

guilpier-code replied:
I do agree that line : 
`d.renewableGeneration.rgModelling = rgAggregated`
is useless.
Line starting with `logs.warning()` could be useless as well, as the `return false` raises a warning as well.
But this last warning is unclear : `renewable-generation-modelling = Unknown property`.
It suggests that property **renewable-generation-modelling** is not a known though it is the value of the property which is not known.
So I left the `logs.warning()` instruction, more clear about that.

guilpier-code replied:
Moreover, I define a **ConvertStringToRenewableGenerationModelling** function, as you said : easier to read.

flomnes commented on src/libs/antares/study/parts/renewable/cluster.h:
```diff
@@ -38,6 +38,15 @@ namespace Antares
 {
 namespace Data
 {
+
+enum RenewableModulation
+{
+    renewableModulationCost = 0,
+    renewableModulationCapacity,
+    renewableMinGenModulation,
+    renewableModulationMax
+};
```
Not sure we want to implement modulation now. To be discussed with @JMJ-rte and @PPlessiez.

guilpier-code replied:
I am afraid that modulation are already fully developed in UI...

flomnes replied:
We may disable it in release 8.1, then enable it in release 8.2.

Keep data structures, disable them by commenting as few lines as possible and marking them as such with a specific keyword.

flomnes commented on src/libs/antares/study/fwd.cpp:
```diff
@@ -356,6 +356,36 @@ const char* NumberOfCoresModeToCString(NumberOfCoresMode ucores)
+    case rgClusters:
+        return "clusters"; // using renewable clusters
+    case ucUnknown:
+        return "";
```
Wrong enum value

flomnes commented on src/ui/simulator/application/main/build/scenario-builder.cpp:
```diff
@@ -204,6 +205,46 @@ class thermalScBuilderGrid : public basicScBuilderGrid
+    {
+    }
+
+    void create()
```
`override`

flomnes commented on src/ui/simulator/application/main/build/scenario-builder.cpp:
```diff
@@ -204,6 +205,46 @@ class thermalScBuilderGrid : public basicScBuilderGrid
+    {
+        grid_ = new DatagridType(page_.first, renderer_);
+    }
+    void addToNotebook()
```
`override`

flomnes commented on src/ui/simulator/windows/renewables/panel.cpp:
```diff
@@ -0,0 +1,214 @@
+Panel::Panel(Component::Notebook* parent) : Component::Panel(parent)
+ ,
+ // pageThermalTimeSeries(nullptr),
+ // pageThermalPrepro(nullptr),
```
Remove

flomnes commented on src/ui/simulator/windows/renewables/panel.h:
```diff
@@ -0,0 +1,105 @@
+public:
+    // gp : to be adapted
+    //! The page related to the cluster's timeseries
+    Component::Notebook::Page* pageThermalTimeSeries;
```
Rename to `pageRenewableTimeSeries`

flomnes commented on src/ui/simulator/windows/renewables/panel.cpp:
```diff
@@ -0,0 +1,214 @@
+          wxT("TS generator"));
+        */
+        // Time Series
+        pageThermalTimeSeries = subbook->add(
```
Rename to `pageRenewableTimeSeries`

flomnes reviewed (commented):
For the renewable panel, there is a lot of code duplication. I'd consider having a common class, then re-implementations when there are differences (properties, etc.).

guilpier-code committed:
UI : taking into account unit count & nominal capacity in RES clusters main panel

flomnes committed:
Merge pull request #284 from AntaresSimulatorTeam/feature/ui-unitCount-and-nominal-capacity

UI : take unit count and nominal capacity into account in RES clusters main panel

flomnes commented on src/ui/simulator/windows/renewables/cluster.cpp:
```diff
@@ -0,0 +1,155 @@
+      new Component::Datagrid::Component(this, 
+        new Component::Datagrid::Renderer::RenewableClusterCommonModulation(this, notifier)),
+        1,
+        wxALL | wxEXPAND);
```
We may comment these lines to disable modulation, at least temporarily.

However, keep class & function definitions.

flomnes committed:
De-duplicate clusterRename.

flomnes commented on src/libs/antares/study/study.cpp:
```diff
@@ -1154,6 +1154,77 @@ bool Study::thermalClusterRename(Cluster* cluster, ClusterName newName, bool)
+        RenewableCluster* found;
+        if ((found = area.renewable.list.find(newID)))
+        {
+            if (found->name() != newName)
+            {
+                area.invalidateJIT = true;
+                found->setName(newName);
+                return true;
+            }
+            return false;
+        }
+    }
+
+    // gp : to be updated with renewable clusters
+    // We must have the scenario rules to rename properly the cluster
+    scenarioRulesLoadIfNotAvailable();
+
+    // We will temporary override the id of the area in order to have to
+    // the new name in the archive
+    // Otherwise the values associated to the area will be lost.
+    logs.info() << "  renaming renewable cluster '" << cluster->name() << "' into '" << newName
+        << "'";
+ 
+    area.invalidateJIT = true;
+
+    bool ret = true;
+
+    // Archiving data
+    {
+        ret = area.renewable.list.rename(cluster->id(), newName);
+        area.renewable.prepareAreaWideIndexes();
+        ScenarioBuilderUpdater updaterSB(*this);
+    }
+
+    if (uiinfo)
+        uiinfo->reloadAll();
+
+    return ret;
+}
+
```
See PR #287 for de-duplication. To be merged into this branch.

flomnes commented on src/ui/simulator/toolbox/components/datagrid/renderer/area/renewable-modulation.cpp:
```diff
@@ -0,0 +1,223 @@
+{
+namespace Renderer
+{
+RenewableClusterCommonModulation::RenewableClusterCommonModulation(
+  wxWindow* control,
+  Toolbox::InputSelector::RenewableCluster* notifier) :
+ Renderer::Matrix<>(control), pCluster(nullptr)
```
Keep data structures for modulation. Remove related panel from the GUI for now.

flomnes commented on src/ui/simulator/toolbox/components/datagrid/renderer/area/renewable.areasummary.cpp:
```diff
@@ -0,0 +1,256 @@
+    case 2:
+        return wxString() << cluster->unitCount;
+    case 3:
+        return DoubleToWxString(cluster->nominalCapacity);
```
Add installed capacity, same as for Thermal.

flomnes commented on src/ui/simulator/toolbox/components/datagrid/renderer/area/renewable.areasummary.cpp:
```diff
@@ -0,0 +1,256 @@
+    {
+    case 0:
+        return wxStringFromUTF8(cluster->group());
+    case 1:
+        return cluster->enabled ? wxT("Yes") : wxT("no");
```
Yes / **N**o for consistency.

flomnes commented on src/ui/simulator/toolbox/components/datagrid/renderer/area/renewable.areasummary.cpp:
```diff
@@ -0,0 +1,256 @@
+
+IRenderer::CellStyle RenewableClusterSummarySingleArea::cellStyle(int col, int row) const
+{
+    return (col > 0 and Math::Zero(cellNumericValue(col, row)))
+             ? IRenderer::cellStyleDefaultDisabled
+             : (col == 1 || col == 2) ? IRenderer::cellStyleConstraintWeight
+                                      : IRenderer::cellStyleDefault;
```
Avoid nested ternary expressions : prefer `if..else` for clarity.

flomnes commented on src/ui/simulator/toolbox/components/datagrid/renderer/area/renewable.areasummary.cpp:
```diff
@@ -0,0 +1,256 @@
+    }
+};
+
+template<class CheckT>
+static bool UpdateUnsignedLong(uint& value, const String& str)
+{
+    uint l;
+    if (str.to(l))
+    {
+        if (value != l and CheckT::Validate(l))
+        {
+            value = l;
+            MarkTheStudyAsModified();
+            OnInspectorRefresh(nullptr);
+            return true;
+        }
+    }
+    return false;
+}
```
It would be nice if class CheckT had this function, since it is generic *ie* not specific to renewable area summary.

It would need a callback to call when a valid expression is provided.

```cpp
template<class T, class predicateT>
checkAndValidate(T value, predicateT callback)
{
    T l;
    if (str.to<T>(l))
    {
        if (value != l and CheckT::Validate(l))
        {
            value = l;
            callback();
            return true;
        }
    }
    return false;
}

void cb() {
            MarkTheStudyAsModified();
            OnInspectorRefresh(nullptr);
}

checkAndValidate<uint, void(*)(void)>(value, cb);
```

flomnes commented on src/ui/simulator/toolbox/components/datagrid/renderer/area/renewable.areasummary.cpp:
```diff
@@ -0,0 +1,256 @@
+    return false;
+}
+
+template<class CheckT>
+static bool UpdateDouble(double& value, const String& str)
+{
+    double d;
+    if (str.to(d))
+    {
+        if (not Math::Equals<double>(value, d))
+        {
+            if (CheckT::Validate(d))
+            {
+                value = d;
+                OnInspectorRefresh(nullptr);
+                return true;
+            }
+        }
+    }
+    return false;
+}
+
+template<class CheckT>
+static bool UpdateBool(bool& value, const String& str)
+{
+    bool d;
+    if (str.to(d))
+    {
+        if (value != d)
+        {
+            if (CheckT::Validate(d))
+            {
+                value = d;
```
Same remark as above : use templates and move to generic class. Then use where needed (thermal, renewable, elsewhere,  etc.).

flomnes commented on src/ui/simulator/toolbox/components/datagrid/renderer/area/timeseries.cpp:
```diff
@@ -293,6 +293,111 @@ void TimeSeriesThermalCluster::onStudyClosed()
+        return IRenderer::cellStyleMinMax;
+    // Default
+    return Math::Zero(v)
+        ? ((row % 2) ? IRenderer::cellStyleDefaultAlternateDisabled
+            : IRenderer::cellStyleDefaultDisabled)
+        : ((row % 2) ? IRenderer::cellStyleDefaultAlternate : IRenderer::cellStyleDefault);
```
avoid nested ternary expressions.

flomnes commented on src/ui/simulator/toolbox/components/datagrid/renderer/area/timeseries.h:
```diff
@@ -369,6 +370,86 @@ class TimeSeriesThermalCluster final : public Renderer::Matrix<double>
     virtual void onStudyClosed();
 };
 
+/*  
+*   --------------------
+*   RENEWABLE CLUSTERS
+*   --------------------
+*/
+// gp : apart from the constructor, the 2 clusters time series classes are identical !
+// gp : To be factored
+// gp : ==> after a merge from feature/renewable, we change the parent of this class
+// gp : then not sure aove comment is still relevant 
+class TimeSeriesRenewableCluster final : public Renderer::Matrix<double>
+{
+public:
+    typedef Renderer::Matrix<double> AncestorType;
+
+public:
+    TimeSeriesRenewableCluster(wxWindow* control, Toolbox::InputSelector::RenewableCluster* notifier);
```
This class is very similar to its thermal counterpart. Please refactor.

flomnes commented on src/ui/simulator/toolbox/components/datagrid/renderer/area/renewable-modulation.h:
```diff
@@ -0,0 +1,113 @@
+        return Date::hourly;
+    }
+
+protected:
+    virtual void internalRenewableClusterChanged(Antares::Data::RenewableCluster* cluster);
```
Why is this function marked virtual ?

flomnes committed:
Internalize cluster type

flomnes committed:
Use dynamic_cast to determine cluster type

Instead of providing it at construction time, use dynamic_cast to
determine cluster type at runtime.

Save memory space by not storing redundant information.

This is a partial revert of the previous commit.

flomnes committed:
Small fix for thermal clusters

flomnes committed:
Merge pull request #287 from AntaresSimulatorTeam/fix/de-duplicate-cluster-rename

De-duplicate clusterRename

flomnes committed:
UI : rename ENR x -> Other RES x

flomnes committed:
Disable renewable modulation

flomnes committed:
Delete unused source and header

flomnes committed:
Merge pull request #292 from AntaresSimulatorTeam/fix/rename-res

UI : rename ENR x -> Other RES x

flomnes committed:
Merge pull request #293 from AntaresSimulatorTeam/fix/disable-res-modulation

Disable renewable modulation

guilpier-code committed:
UI - Renewable cluster main panel : Changing units number

Allow UI user to change the units number in Common/RENEWABE CLUSTER panel

flomnes committed:
UI : Add installed capacity for renewable clusters

Main panel & inspector

flomnes committed:
Merge pull request #294 from AntaresSimulatorTeam/fix/ui-changing-units-number

UI Allow change of units number

guilpier-code committed:
UI - renewable clusters : allowing cloning clusters (main clusters panel) (#296)

Co-authored-by: Guillaume PIERRE <guillaume.pierre@rte-france.com>

guilpier-code committed:
UI : allowing to delete all renewable clusters of an area

flomnes committed:
Refactor Value structs

Remove duplicate structs, share functions between thermal & renewable clusters.

flomnes committed:
UI : Add ts mode for renewable clusters

guilpier-code committed:
UI - renewable clusters : remove market bid trace ("€/MWh") in front of each cluster on main cluster panel

flomnes committed:
PR remarks

flomnes committed:
Merge pull request #301 from AntaresSimulatorTeam/feature/ui-remove-market-bid-on-cluster-end-line

UI : remove "€/MWh" in front of each cluster

flomnes committed:
Merge remote-tracking branch 'http/feature/renewable' into HEAD

flomnes committed:
Fix build - duplicate struct declaration

flomnes committed:
Merge branch 'feature/ui-res-installed-cap' into feature/ui-res-ts-interpretation

flomnes committed:
Merge pull request #295 from AntaresSimulatorTeam/feature/ui-res-installed-cap

UI : Add installed capacity for renewable clusters

flomnes committed:
Merge pull request #299 from AntaresSimulatorTeam/feature/ui-allow-delete-all-clusters

UI : allowing to delete all renewable clusters of an area

flomnes committed:
Merge remote-tracking branch 'http/feature/merge-from-feature-renewable' into feature/ui-res-ts-interpretation

flomnes committed:
Fix cluster duplication

tsMode was not duplicated

guilpier-code committed:
UI : more specific event when changing renewable generation modeling parameter

flomnes committed:
Refactor button creation

TODO : same refactor for other buttons

flomnes committed:
Add "Ren. clusters" button

flomnes committed:
Enable details-res button

guilpier-code committed:
Reading renewable modelling parameter : enhancement of readability

guilpier-code committed:
Scenario builder : for thermal and renewable grid classes, overriding methods of base class methods

guilpier-code committed:
renewable cluster => area summary : "No" (not "no") if cluster disabled

guilpier-code committed:
renewable cluster => area summary : replacing ternary a ? b : c nested expressions with if..else, more clear

guilpier-code committed:
Time series grids (specially for clusters) : replacing nested ternary operators by if..else statements, for more clarity

guilpier-code committed:
Renewable clusters main panel : some cleaning

guilpier-code committed:
Renewable clusters main panel : more cleaning

flomnes committed:
Merge pull request #302 from AntaresSimulatorTeam/feature/ui-res-ts-interpretation

UI : Add ts mode for renewable clusters

flomnes committed:
Merge pull request #307 from AntaresSimulatorTeam/fix/ui-more-specific-event-for-switch-renewable

UI : more specific event for switch renewable

flomnes committed:
Merge pull request #309 from AntaresSimulatorTeam/feature/output-details-res

Add renewable detail to output
